python/ql/lib/semmle/python/Concepts.qll

```diff
@@ -1231,6 +1246,104 @@ module Http {
          * Gets the argument, if any, specifying the cookie value.
          */
         abstract DataFlow::Node getValueArg();
+
+        /**
+         * Holds if the `Secure` flag of the cookie is known to have a value of `b`.
+         */
+        predicate hasSecureFlag(boolean b) {
+          exists(this.getHeaderArg()) and
+          (
+            exists(StringLiteral sl |
+              sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") and
+              TaintTracking::localTaint(DataFlow::exprNode(sl), this.getHeaderArg()) and
+              b = true
+            )
+            or
+            exists(StringLiteral sl |
+              not sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") and
+              DataFlow::localFlow(DataFlow::exprNode(sl), this.getHeaderArg()) and
+              b = false
+            )
+          )
+        }
```

Comment:
```suggestion
        predicate hasSecureFlag(boolean b) {
          exists(StringLiteral sl |
            TaintTracking::localTaint(DataFlow::exprNode(sl), this.getHeaderArg())
          |
            if sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") then b = true else b = false
          )
        }
```
I think this reads a bit easier: If there is in fact a literal defining the header argument, then we check it against the regexp and know the flag accordingly.
(The existence of `this.getHeaderArg()` is implied by the `localTaint` demand).

Reply:
This does have a functional difference - currently, the `true` case uses `TaintTracking`, which allows for `"; Secure"` to be concatenated onto the cookie header, but the `false` case requires a `DataFlow` ensuring that we only report the flag as being not set when we know the exact string that is used. (in particular, we wouldn't report an alert in a case like `resp['Set-Cookie'] = "key=value; " + "Secure; " + "HttpOnly")`

Reply:
Ah! I overlooked that, that is nice, actually :-)

Reply:
So, let me try again :-)
```suggestion
        predicate hasSecureFlag(boolean b) {
          exists(StringLiteral sl |
            // `sl` is likely a substring of the header
            TaintTracking::localTaint(DataFlow::exprNode(sl), this.getHeaderArg()) and
            sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") and
            b = true
            or
            // `sl` is the entire header
            DataFlow::localFlow(DataFlow::exprNode(sl), this.getHeaderArg()) and
            not sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") and
            b = false
          )
        }
```
So there is a literal in the code; either it gets to the header approximately and matches the regex, in which case the header now probably contains the string "; Secure", or it gets there exactly and does not match, in which case the header definitely does not contain the string "; Secure".